### PR TITLE
Use slugs for attachments urls, add badge for direct writeup downloads

### LINKED ISSUE #2035: Add API/CLI endpoint for attachments
Issues with current implementation of attachments (as per #1983)
- does not support static links (tedious to update existing permalinks)
- no easy way to mass-import

For that reason, using the assessment "direct-download" feature (#2020) is somewhat more convenient for instructors
- only course name changes, easy to update links
- can mass import assessments

Potential solution
- Add an endpoint which uploads an attachment and **returns the attachment link** -> easy to generate HTML lists
- Rename "attachments" to "course materials"

Also possible: allow the specification of a canonical name

Note: An ad-hoc workaround was available in #2020 #2021 but this would be a longer-term solution.

### PR DESCRIPTION
**Pull request by damianhxy**
## Description

- Generate slugged URLs for attachments
- Allow direct download of writeups from course landing page
- Fix highlighting on docs

## Motivation and Context

Some classes might use assessments as "placeholders" to distribute course materials (via writeups), as compared to attachments, since assessments can be easily bulk imported and the URL is "static" (only need to change course name between semesters). This PR further facilitates that workflow by allowing students to quickly download an assessment writeup from a course landing page by clicking on a badge. This is also useful for students to view writeups of normal assessments quickly.

This PR also switches to slugged URLs for attachments, i.e. the URL is generated from the attachment name. This would be helpful down the road if a way were added to mass import attachments since attachments would become a viable replacement for assessments.

This PR also fixes a docs syntax highlighting issue introduced by #2131.

Resolves #2035

### Slugged Attachments
- Create a new course / assessment attachment -> observe that the URL now uses the attachment name as a slug
- Update the name of an attachment -> observe that the slug updates accordingly
- Cause a name collision -> id appended to ensure uniqueness

### Writeup badges
Screenshot 2024-04-25 at 01 58 14

Badge appears only for assessments with writeups. Clicking on the badge brings you to the writeup.

### Docs highlighting
Run `mkdocs serve` locally and ensure that the bash commands on e.g. the docker compose installation page now render correctly.

**Before**
Screenshot 2024-04-20 at 15 42 22

**After**
Screenshot 2024-04-20 at 15 42 12

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have run rubocop and erblint for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://docs.autolabproject.com/)
- [ ] I have updated the documentation accordingly, included in this PR

### CODE REVIEW

File: app/models/attachment.rb [hunk 1]
```diff
@@ -44,6 +46,11 @@ def file=(upload)
     self.mime_type = upload.content_type
   end
 
+  # Regenerate slug whenever the name changes
+  def should_generate_new_friendly_id?
+    name_changed? && name.present?
+  end
```
[the remark this answers is not in the document]
Comment (damianhxy) [reply app/models/attachment.rb:52]: Collisions are handled by the FriendlyId gem that we're using

**Comment by KesterTan** [comment]
Not sure if this is desired but when for existing attachments, the url does not include the attachment name, it only includes after I rename the attachment or upon creating a new attachment

**Review by KesterTan — COMMENTED** [review]
Not sure if this is desired but when for existing attachments, the url does not include the attachment name, it only includes after I rename the attachment or upon creating a new attachment

File: app/views/assessments/index.html.erb [hunk 2]
```diff
@@ -165,6 +165,9 @@
                         <span class="new badge blue darken-4" data-badge-caption="unreleased"></span>
                       <% end %>
                     <% end %>
+                    <% if asmt.has_writeup? %>
+                      <span class="new badge orange darken-4" data-badge-caption="writeup" data-url="<%= writeup_course_assessment_path(@course, asmt) %>"></span>
```
Comment (KesterTan) [inline app/views/assessments/index.html.erb:169]: Would prefer if there was some gap between the writeup flag and the unreleased flag. Now they look too close together.

**Comment by KesterTan** [comment]
Also, is there a way to shorten the url in the case of duplicates, "test778-cf1a1964-f358-4cff-8136-8c9791b40ec6" this seems a little too long. Would it make sense if we generated a random string usingSecureRandom and took the difference to get a shortened slug that is still unique.

File: app/views/assessments/index.html.erb (hunk 2 above)
Comment (damianhxy) [reply app/views/assessments/index.html.erb:169]: The margin was removed in #2016, but I can add it back

**Comment by damianhxy** [comment]
> Also, is there a way to shorten the url in the case of duplicates, "test778-cf1a1964-f358-4cff-8136-8c9791b40ec6" this seems a little too long. Would it make sense if we generated a random string usingSecureRandom and took the difference to get a shortened slug that is still unique.

Modified the code to tiebreak by appending the attachment's id

File: app/controllers/attachments_controller.rb [hunk 3]
```diff
@@ -96,13 +96,9 @@ def update
...
-      error_msg = "Attachment update failed:"
-      if !@attachment.valid?
-        @attachment.errors.full_messages.each do |msg|
-          error_msg += "<br>#{msg}"
-        end
-      else
-        error_msg += "<br>Unknown error"
+      error_msg = "Attachment update failed."
+      @attachment.errors.full_messages.each do |msg|
+        error_msg += "<br>#{msg}"
```
[the remark this answers is not in the document]
Comment (damianhxy) [reply app/controllers/attachments_controller.rb:101]: The current logic suffices

**Comment by damianhxy** [comment]
> Not sure if this is desired but when for existing attachments, the url does not include the attachment name, it only includes after I rename the attachment or upon creating a new attachment

That was intended and IMO it's fine, attachments don't seem to be widely used yet, and this change is more to facilitate attachments in future courses.

According to the documentation, you'd have to do something similar to `User.find_each(&:save)`. If you feel like it's important to generate slugs retroactively, I could put this in the migration

**Review by KesterTan — CHANGES_REQUESTED** [review]
Small nit: I think some padding top: maybe  .date p { padding-top: 4px} would be even more perfect.

**Review by KesterTan — APPROVED** [review]
LGTM, everything works as expected (slugs look nice and clean gj!)

**Pull request merged**